### PR TITLE
Make the expression evaluator support heterogeneous results

### PR DESCRIPTION
**Pull request by jachris**
Functions can now return values of different types for the same input types. For example, `x.otherwise(y)` no longer requires that `x` has the same type as `y`.

Metadata such as `@name` can now be set to a dynamically computed value that does not have to be a constant. For example, if the field `event_name` should be used as the event name, `@name = event_name` now correctly assigns the events their name instead of using the first value.

### CODE REVIEW

File: libtenzir/builtins/functions/otherwise.cpp [hunk 1]
```diff
@@ -30,37 +30,62 @@ class otherwise final : public function_plugin {
...
+                return ps.array->IsValid(i);
+              }
+              return not last_valid;
+            });
+            if (last_valid != valid) {
+              if (valid) {
+                parts.push_back(ps.slice(begin, i));
+              } else {
+                parts.push_back(fs.slice(begin, i));
+              }
```
[the remark this answers is not in the document]
Comment (jachris) [reply libtenzir/builtins/functions/otherwise.cpp]: Thanks!

File: libtenzir/include/tenzir/multi_series.hpp [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+      }
+      row -= part.length();
+    }
+    TENZIR_UNREACHABLE();
+  }
+
+  std::vector<series> parts_;
+};
+
+// TODO: Lifetime??
```
Comment (dominiklohmann) [inline libtenzir/include/tenzir/multi_series.hpp:111]: This looks like a TODO that should be addressed.

File: libtenzir/include/tenzir/multi_series.hpp [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+                detail::function_view<auto(std::span<series>)->multi_series> f)
+  -> multi_series;
+
+auto map_series(multi_series x,
+                detail::function_view<auto(series)->multi_series> f)
+  -> multi_series;
+
+auto map_series(multi_series x, multi_series y,
+                detail::function_view<auto(series, series)->multi_series> f)
+  -> multi_series;
```
Comment (dominiklohmann) [inline libtenzir/include/tenzir/multi_series.hpp:138]: I think this could definitely use documentation.

File: libtenzir/src/multi_series.cpp [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+//
+// SPDX-FileCopyrightText: (c) 2024 The Tenzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "tenzir/multi_series.hpp"
+
+namespace tenzir {
+
+auto split_multi_series(std::span<const multi_series> input,
+                        std::span<series> output) -> generator<std::monostate> {
```
Comment (dominiklohmann) [inline libtenzir/src/multi_series.cpp:14]: The TODOs in this function should at the very least be addressed in assertions.

File: libtenzir/src/tql2/eval.cpp [hunk 5]
```diff
@@ -79,7 +89,8 @@ auto const_eval(const ast::expression& expr, diagnostic_handler& dh)
     auto sp = session_provider::make(dh);
     auto result = evaluator{nullptr, sp.as_session()}.eval(expr);
     TENZIR_ASSERT(result.length() == 1);
-    return materialize(value_at(result.type, *result.array, 0));
+    auto& part = result.part(0);
```
Comment (dominiklohmann) [inline libtenzir/src/tql2/eval.cpp:92]: This should probably assert that there is exactly one part?

File: libtenzir/src/tql2/eval_impl.cpp [hunk 6]
```diff
@@ -7,167 +7,194 @@
...
-    = make_struct_array(length_, nullptr,
-                        std::vector(field_names.begin(), field_names.end()),
-                        std::vector(field_arrays.begin(), field_arrays.end()));
-  return series{
-    type{record_type{std::vector(field_types.begin(), field_types.end())}},
-    std::move(result),
-  };
+  return map_series(items, [&](std::span<series> items) {
+    auto fields = detail::stable_map<std::string, series>{};
+    for (auto [item, desc] : detail::zip_equal(items, x.items)) {
```
Comment (dominiklohmann) [inline libtenzir/src/tql2/eval_impl.cpp:35]: I find this code too hard to follow. Why is `x.items` containing descriptions?

File: libtenzir/src/tql2/set.cpp [hunk 7]
```diff
@@ -240,16 +308,48 @@ auto set_operator::operator()(generator<table_slice> input,
...
+    auto values = std::vector<multi_series>{};
     for (auto& assignment : assignments_) {
-      auto right = eval(assignment.right, slice, ctrl.diagnostics());
-      result = assign(assignment.left, right, result, ctrl.diagnostics());
+      values.push_back(eval(assignment.right, slice, ctrl.diagnostics()));
+    }
+    // After we know all the multi series values on the right, we can split the
+    // input table slice and perform the actual assignment.
+    auto begin = int64_t{0};
+    for (auto values_slice : split_multi_series(values)) {
```
Comment (dominiklohmann) [inline libtenzir/src/tql2/set.cpp:325]: Did you consider pushing ordering assumptions into this operator (or perhaps more generally into the new abstractions)? I assume this could make a pretty big difference in performance.

**Review by dominiklohmann — COMMENTED** [review]
I went through all the code once and did some initial tests; things seem to be running well. I'll continue testing for now, but here's an initial round of feedback. I love seeing this come to live. Thanks for minimizing the complexity for function and operator authors, too.

**Review by dominiklohmann — APPROVED** [review]
I've gone through all the code in detail now, and I've tested various combinations of this. Besides the already existing comments I think this is ready to go.

Thank you for pushing this forward!

File: libtenzir/include/tenzir/multi_series.hpp (hunk 2 above)
Comment (jachris) [reply libtenzir/include/tenzir/multi_series.hpp]: I tried out a few things like marker types, but ultimately this is just down to the lack of lifetime annotations. I just removed the comment.

Comment (dominiklohmann) [reply libtenzir/include/tenzir/multi_series.hpp]: Thanks for the explanation. Let's keep it as-is then.
